src/internals.jl

```diff
@@ -0,0 +1,200 @@
+module Internals
+
+using ChunkSplitters: Split, Consecutive, RoundRobin
+import ChunkSplitters: chunk_indices, chunk, is_chunkable
+import Base: iterate, length, eltype, enumerate, firstindex, lastindex, getindex, eachindex
+
+abstract type Constraint end
+struct FixedCount <: Constraint end
+struct FixedSize <: Constraint end
+
+abstract type ReturnType end
+struct ReturnIndices <: ReturnType end
+struct ReturnViews <: ReturnType end
+
+struct ChunksIterator{T,C<:Constraint,S<:Split,R<:ReturnType}
+    collection::T
+    n::Int
+    size::Int
+end
+
+function ChunksIterator(s::Split, r::ReturnType; collection, n=nothing, size=nothing, minchunksize=nothing)
+    is_chunkable(collection) || err_not_chunkable(collection)
+    isnothing(n) && isnothing(size) && err_missing_input()
+    !isnothing(size) && !isnothing(n) && err_mutually_exclusive("size", "n")
+    C, n, size = _set_C_n_size(collection, n, size, minchunksize)
+    return ChunksIterator{typeof(collection),C,typeof(s),typeof(r)}(collection, n, size)
+end
+
+# public API
+function chunk_indices(collection;
+    n::Union{Nothing,Integer}=nothing,
+    size::Union{Nothing,Integer}=nothing,
+    split::Split=Consecutive(),
+    minchunksize::Union{Nothing,Integer}=nothing,
+)
+    return ChunksIterator(split, ReturnIndices(); collection, n, size, minchunksize)
+end
+
+# public API
+function chunk(collection;
```

Comment:
Why not `chunks`? It returns multiple `chunk`s. Or is this a verb and not a noun? To me `chunks(...)::Chunks` and `index_chunks(...)::IndexChunks` would feel more natural.

Reply:
> Why not chunks? It returns multiple chunks. Or is this a verb and not a noun? To me chunks(...)::Chunks and index_chunks(...)::IndexChunks would feel more natural.

It's supposed to be a verb similar to `enumerate`, `partition`, and `zip`. However, we also have `pairs` and `eachindex`, so we could also use a noun instead. Frankly, I don't have a super strong opinion here. It has crossed my mind that `chunk` might be more likely to be interpreted as a noun and might lead to confusion.

Reply:
I find more natural "chunks", but since I don't use this word in any other context in life I can get used to any alternative.

Reply:
I went with @fredrikekre's suggestion.